src/libmongoc/src/mongoc/mongoc-write-command.c

```diff
@@ -1355,6 +1357,22 @@ _mongoc_write_result_merge (mongoc_write_result_t *result,   /* IN */
       result->n_writeConcernErrors++;
    }
 
+   /* If a server error ocurred, then append the raw response to the
+    * error_replies array. */
+   if (result->failed || result->n_writeConcernErrors ||
+       _mongoc_error_is_server (&result->error)) {
```

Comment:
`result->n_writeConcernErrors` is a running count in `mongoc_write_result_t`. The `mongoc_write_result_t` is the result of merging possibly many server replies. If `n_writeConcernErrors > 0`, that does not imply the current reply is an error.

Suggest adding `bool` variables to track existence of "writeErrors" and "writeConcernError" in the if blocks above, and using `_mongoc_cmd_check_ok` to check for the `'ok':0` case:

```suggestion
   /* If a server reply contains 'writeErrors', 'writeConcernError' or 'ok':0,
    * then append the raw response to the error_replies array. */
   if (has_writeErrors || has_writeConcernError ||
       !_mongoc_cmd_check_ok (
          reply, MONGOC_ERROR_API_VERSION_2, NULL /* error */)) {
```

Reply:
Good catch.